data/preprocess/util.py

```diff
@@ -0,0 +1,180 @@
+    if create_filtered:
+        filtered_size = 0
+        filter_entities = kwargs["filter_entities"]
+        filter_relations = kwargs["filter_relations"]
+        filter_f = open(kwargs["filtered_file"], "w")
+
+    with open(file, "w") as f:
+        for n, t in enumerate(raw_split):
+            write_triple(f, entities, relations, t, S, P, O)
+            if create_sample and n in sample:
+                write_triple(sample_f, entities, relations, t, S, P, O)
+            if create_filtered and t[S] in filter_entities \
+                               and t[O] in filter_entities \
+                               and t[P] in filter_relations:
+                write_triple(filter_f, entities, relations, t, S, P, O)
+                filtered_size += 1
+        if create_filtered:
+            return filtered_size
+
+
+def process_pos_neg_split(
+        entities: dict,
+        relations: dict,
+        pos_file: str,
+        neg_file: str,
+        raw_split: list,
+        order_sop: bool = False,
+        create_filtered: bool = False,
+        **kwargs
+):
+    """From a raw split containing labeled triples, write split files with indexes.
+
+     Optionally, filtered split files can be created.
+
+     """
+    if order_sop:
+        S, P, O = 0, 2, 1
+    else:
+        S, P, O = 0, 1, 2
+
+    # the filtered split cosists of triples from the raw split where both entities
+    # and relation exist in filter_entities/filter_relation
+    if create_filtered:
+        filter_entities = kwargs["filter_entities"]
+        filter_relations = kwargs["filter_relations"]
+        filter_pos_f = open(kwargs["filtered_pos_file"], "w")
+        filter_neg_f = open(kwargs["filtered_neg_file"], "w")
+        filtered_pos_size = 0
+        filtered_neg_size = 0
+
+    pos_size = 0
+    neg_size = 0
+    with open(pos_file, "w") as pos_file, \
+         open(neg_file, "w") as neg_file:
+
+        for n, t in enumerate(raw_split):
+            if int(t[3]) == -1:
+                file_wrapper = neg_file
+                filtered_file_wrapper = filter_neg_f
+                neg_size += 1
+            else:
+                file_wrapper = pos_file
+                filtered_file_wrapper = filter_pos_f
+                pos_size += 1
+            write_triple(file_wrapper, entities, relations, t, S, P, O)
+            if create_filtered and t[S] in filter_entities \
+                               and t[O] in filter_entities \
+                               and t[P] in filter_relations:
+
+                if int(t[3]) == -1:
+                    filtered_neg_size += 1
+                else:
+                    filtered_pos_size += 1
+                write_triple(filtered_file_wrapper, entities, relations, t, S, P, O)
+
+        if create_filtered:
+            return pos_size, filtered_pos_size, neg_size, filtered_neg_size
+        else:
+            return pos_size, neg_size
+
+
+def write_obj_meta(config: dict):
```

Comment:
This could be merged to a method that takes the analyzed dataset and writes entity/relation ids. All methods generally could take a (dataset) config and modify it based on what they do.